internal/core/software_renderer.rs

```diff
@@ -314,12 +314,76 @@ impl Renderer for SoftwareRenderer {
     ) -> usize {
         0
     }
+
     fn text_input_cursor_rect_for_byte_offset(
         &self,
-        _text_input: Pin<&crate::items::TextInput>,
-        _byte_offset: usize,
+        text_input: Pin<&crate::items::TextInput>,
+        byte_offset: usize,
     ) -> LogicalRect {
-        Default::default()
+        let window_adapter = match self.window.upgrade() {
+            Some(window) => window,
+            None => return Default::default(),
+        };
+
+        let scale_factor = ScaleFactor::new(window_adapter.window().scale_factor());
+        let string = text_input.text();
+
+        let font_request = text_input.font_request(&window_adapter);
+        let font = fonts::match_font(&font_request, scale_factor);
+
+        let width: Length<i16, crate::lengths::PhysicalPx> =
+            Length::new((text_input.width().0 * scale_factor.0).round() as i16);
+        let height: Length<i16, crate::lengths::PhysicalPx> =
+            Length::new((text_input.height().0 * scale_factor.0).round() as i16);
```

Comment:
Calling `Length::new()` very likely to introduce bugs.

```suggestion
        let width =
            (text_input.width() * scale_factor).cast();
        let height =
            (text_input.height() * scale_factor).cast();
```